adapter/service/battery/service.go

```diff
@@ -0,0 +1,321 @@
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatteryHealthReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	health, err := s.reporter.BatteryHealthReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery health report: %w", err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtHealthReport, "", health) {
+		return false, nil
+	}
+
+	message := fimpgo.NewIntMessage(
+		EvtHealthReport,
+		s.Name(),
+		health,
+		nil,
+		nil,
+		nil,
+	)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("failed to send battery health report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtHealthReport, "", health)
+
+	return true, nil
+}
+
+// SendBatterySensorReport sends a battery sensor report. Returns true if a report was sent.
+// Depending on a caching and reporting configuration the service might decide to skip a report.
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatterySensorReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	sensor, unit, err := s.reporter.BatterySensorReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery sensor report: %w", err)
+	}
+
+	props := fimpgo.Props{
+		"unit": unit,
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtSensorReport, "", sensor) {
+		return false, nil
+	}
+
+	message := fimpgo.NewFloatMessage(
+		EvtSensorReport,
+		s.Name(),
+		sensor,
+		props,
+		nil,
+		nil,
+	)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("failed to send battery sensor report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtSensorReport, "", sensor)
+
+	return true, nil
+}
+
+// SendBatteryFullReport sends a full battery report. Returns true if a report was sent.
+// Depending on a caching and reporting configuration the service might decide to skip a report.
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatteryFullReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	full, err := s.reporter.BatteryFullReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery full report: %w", err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtBatteryReport, "", full) {
+		return false, nil
+	}
+
+	message := fimpgo.NewObjectMessage(
+		EvtBatteryReport,
+		s.Name(),
+		full,
+		nil,
+		nil,
+		nil,
+	)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("failed to send battery full report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtBatteryReport, "", full)
+
+	return true, nil
+}
+
+func (bar *AlarmReport) StructToStrMap() (strMap map[string]string, err error) {
+	marshal, err := json.Marshal(bar)
```

Comment:
I'd return a map directly instead of marshalling/unmarshalling as those are pretty expensive operations.